gems/sorbet-runtime/lib/types/private/methods/decl_builder.rb

```diff
@@ -15,6 +29,31 @@ class BuilderError < StandardError; end
       end
     end
 
+    private def check_sig_block_is_unset!
+      if T::Private::DeclState.current.active_declaration&.blk
+        raise BuilderError.new(
+          "Cannot add more signature statements after the declaration block."
+        )
+      end
+    end
+
+    # Verify if we're trying to invoke the method outside of a signature block. Notice that we need to check if it's a
+    # proc, because this is valid and would lead to a false positive: `T.type_alias { T.proc.params(a: Integer).void }`
+    private def check_running_inside_block!(method_name)
+      unless @inside_sig_block || decl.mod == T::Private::Methods::PROC_TYPE
+        raise BuilderError.new(
+          "Can't invoke #{method_name} outside of a signature declaration block"
+        )
+      end
+    end
+
+    def deep_clone
+      fresh_builder = self.class.new(decl.mod, decl.raw)
+      fresh_builder.decl = decl.dup
```

Comment:
I can't figure out what I would do differently here, but this code does not spark joy.

I am slightly concerned that in `deep_clone`, we're really not deep cloning that much: `decl.dup` seems to be the only piece of data that we're actually doing anything with, and we're not even dup'ing the pieces of `decl` used to form the new builder.

Also a little concerned that having this here means that we need to be careful about things we add to `Declaration`; we now need to decide whether we need to deal with them in this function -- and we need to remember to do that considering in the first place.

Reply:
We can inline this in `run_builder` to match previous implementation. That might get rid of part of the code smells.

Another option is changing the `initialize` of `DeclBuilder` to take in a full `Declaration` Struct, (`decl` field). I'm not sure if it'll work with other callsites but just by looking at this usage, accepting a `Declaration` conceptually makes sense to me.